packages/utils/block_time_calculator.go

```diff
@@ -0,0 +1,93 @@
+//go:generate sh -c "mockery -inpkg -name Clock -print > file.tmp && mv file.tmp clock_mock.go"
+
+package utils
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// BlockTimeCalculator calculating block generation time
+type BlockTimeCalculator struct {
+	clock Clock
+
+	firstBlockTime      time.Time
+	blockGenerationTime time.Duration
+	blocksGap           time.Duration
+
+	nodesCount int64
+}
+
+type blockGenerationState struct {
+	start    time.Time
+	duration time.Duration
+
+	nodePosition int
+}
+
+var TimeError = errors.New("current time before first block")
+
+func NewBlockTimeCalculator(firstBlockTime time.Time,
+	blockGenerationTime, blocksGap time.Duration,
+	nodesCount int64,
+) BlockTimeCalculator {
+	return BlockTimeCalculator{
+		firstBlockTime:      firstBlockTime,
+		blockGenerationTime: blockGenerationTime,
+		blocksGap:           blocksGap,
+		nodesCount:          nodesCount,
+	}
+}
+
+func (btc *BlockTimeCalculator) TimeToGenerate(nodePosition int64) (bool, error) {
+	bgs, err := btc.countBlockTime(btc.clock.Now())
+	if err != nil {
+		return false, err
+	}
+
+	return int64(bgs.nodePosition) == nodePosition, nil
```

Comment:
may be better to change bgs.nodePosition type to int64, than converting it every time?